File: backend/test/unicode_round_trip.test.ts
/** Proves Unicode survives API parsing, validation, storage and response JSON. */

import assert from "node:assert/strict";
import test from "node:test";

import { FolooApplication } from "../src/application/foloo_application.js";
import { createRouter } from "../src/transport/router.js";
import { apiEvent, MemoryRepository } from "./helpers.js";

const unicodeText =
  "MÉXICO · José Álvarez · León, Guanajuato · Niñez y pingüino · " +
  "¿Información? · ¡Hola! · España · São Paulo";
const eventPayload = {
  id: "b9174606-f614-4f11-975b-d5a777427605",
  name: unicodeText,
  startsAt: "2026-09-15T10:00:00Z",
  endsAt: "2026-09-16T10:00:00Z",
};

test("RNF-06 request, validation, repository and response preserve Unicode", async () => {
  const repository = new MemoryRepository();
  const router = createRouter(new FolooApplication(repository));

  const createResponse = await router(
    apiEvent({
      method: "POST",
      path: "/v1/events",
      subject: "unicode-owner",
      idempotencyKey: "unicode-event-request",
      body: eventPayload,
    }),
  );

  assert.equal(createResponse.statusCode, 201);
  assert.equal(
    (JSON.parse(createResponse.body ?? "{}") as { data: { name: string } }).data
      .name,
    unicodeText,
  );
  assert.equal(
    (
      repository.eventsByWorkspace.get("workspace-unicode-owner")?.[0] as {
        name: string;
      }
    ).name,
    unicodeText,
  );

  const listResponse = await router(
    apiEvent({ path: "/v1/events", subject: "unicode-owner" }),
  );
  assert.equal(
    (
      JSON.parse(listResponse.body ?? "{}") as {
        data: Array<{ name: string }>;
      }
    ).data[0]?.name,
    unicodeText,
  );
  assert.equal(
    listResponse.headers?.["content-type"],
    "application/json; charset=utf-8",
  );
});

test("RNF-06 base64 API Gateway bodies decode explicitly as UTF-8", async () => {
  const repository = new MemoryRepository();
  const router = createRouter(new FolooApplication(repository));
  const event = apiEvent({
    method: "POST",
    path: "/v1/events",
    subject: "unicode-owner",
    idempotencyKey: "unicode-base64-request",
  });
  event.body = Buffer.from(JSON.stringify(eventPayload), "utf8").toString(
    "base64",
  );
  event.isBase64Encoded = true;

  const response = await router(event);

  assert.equal(response.statusCode, 201);
  assert.equal(
    (JSON.parse(response.body ?? "{}") as { data: { name: string } }).data.name,
    unicodeText,
  );
});

File: app/test/unicode_round_trip_test.dart
// Verifies Unicode survives Foloo's local, JSON, HTTP and sync boundaries.
// These tests protect all business text carried by the shared persistence and
// transport layers rather than special-casing Spanish characters.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:foloo/data/local/app_database.dart';
import 'package:foloo/data/repositories/local_repositories.dart';
import 'package:foloo/models/app_event.dart';
import 'package:foloo/sync/foloo_api_client.dart';
import 'package:foloo/sync/sync_engine.dart';
import 'package:foloo/sync/sync_models.dart';
import 'package:foloo/sync/sync_store.dart';

const _unicodeText =
    'MÉXICO · José Álvarez · León, Guanajuato · Niñez y pingüino · '
    '¿Información? · ¡Hola! · España · São Paulo';
const _owner = 'unicode-owner';
const _eventId = 'b9174606-f614-4f11-975b-d5a777427605';

class _Session implements SyncSessionProvider {
  @override
  Future<String?> accessTokenFor(String ownerSub) async => 'test-token';
}

class _RoundTripApi implements SyncApi {
  final remoteEvents = <Map<String, Object?>>[];

  @override
  Future<SyncResponse> send(String accessToken, SyncRequest request) async {
    if (request.method == 'POST' && request.path == '/v1/events') {
      final encoded = jsonEncode(request.body);
      final event = (jsonDecode(encoded) as Map).cast<String, Object?>();
      remoteEvents.add({...event, 'revision': 1});
      return SyncResponse(statusCode: 201, data: {'data': event});
    }
    if (request.path == '/v1/profile') {
      return const SyncResponse(statusCode: 200, data: {'data': null});
    }
    if (request.path == '/v1/events') {
      return SyncResponse(statusCode: 200, data: {'data': remoteEvents});
    }
    if (request.path == '/v1/leads') {
      return const SyncResponse(statusCode: 200, data: {'data': []});
    }
    if (request.path == '/v1/content') {
      return const SyncResponse(statusCode: 200, data: {'data': []});
    }
    throw StateError('Unexpected request: ${request.method} ${request.path}');
  }
}

void main() {
  test('RNF-06 Unicode survives Drift, outbox, API and pull', () async {
    final database = AppDatabase(NativeDatabase.memory());
    addTearDown(database.close);
    final events = EventRepository(database);
    final store = SyncStore(database);

    await events.save(
      _owner,
      AppEvent(
        id: _eventId,
        name: _unicodeText,
        startsOn: DateTime.utc(2026, 9, 15),
        endsOn: DateTime.utc(2026, 9, 16),
      ),
    );

    expect((await events.list(_owner)).single.name, _unicodeText);
    final operation = (await store.all(_owner)).single;
    final payload = jsonDecode(operation.payloadJson) as Map<String, dynamic>;
    expect(payload['name'], _unicodeText);
    expect(jsonDecode(jsonEncode(payload))['name'], _unicodeText);

    final api = _RoundTripApi();
    await SyncEngine(store, api, _Session()).synchronize(_owner);

    expect(api.remoteEvents.single['name'], _unicodeText);
    expect((await events.list(_owner)).single.name, _unicodeText);
  });

  test('RNF-06 HTTP transport writes and decodes JSON as UTF-8', () async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    addTearDown(server.close);
    final receivedBytes = Completer<List<int>>();
    server.listen((request) async {
      final bytes = await request.fold<List<int>>(
        <int>[],
        (all, chunk) => all..addAll(chunk),
      );
      receivedBytes.complete(bytes);
      request.response.headers.contentType = ContentType(
        'application',
        'json',
        charset: 'utf-8',
      );
      request.response.add(
        utf8.encode(
          jsonEncode({
            'data': {'name': _unicodeText},
          }),
        ),
      );
      await request.response.close();
    });

    final body = jsonEncode({'name': _unicodeText});
    final response = await IoSyncHttpTransport().send(
      uri: Uri.parse('http://127.0.0.1:${server.port}/unicode'),
      method: 'POST',
      headers: const {'content-type': 'application/json; charset=utf-8'},
      body: body,
      timeout: const Duration(seconds: 5),
    );

    expect(await receivedBytes.future, utf8.encode(body));
    expect((response.data as Map)['data']['name'], _unicodeText);
  });
}
